File: 卡码/代码随想录/Python/哈希/08-四数之和/LeetCode_0018.py
#18. 四数之和
from typing import *

class Solution:
	def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
		res=[]
		nums.sort()
		for k in range(len(nums)-3):
			if nums[k]>target and nums[k]>=0:
				break
			if k>0 and nums[k]==nums[k-1]:
				continue
			for i in range(k+1,len(nums)-2):
				if nums[k]+nums[i]>target and nums[i]>=0:
					break
				if i>k+1 and nums[i]==nums[i-1]:
					continue
				left,right=i+1,len(nums)-1
				while left<right:
					n1,n2=nums[left],nums[right]
					if nums[k]+nums[i]+n1+n2==target:
						res.append([nums[k],nums[i],n1,n2])
						while left<right and nums[left]==n1:
							left+=1
						while left<right and nums[right]==n2:
							right-=1
					elif nums[k]+nums[i]+n1+n2<target:
						left+=1
					else:
						right-=1
		return res

File: 卡码/代码随想录/C++/哈希/08-四数之和/LeetCode_0018.cpp
//18. 四数之和
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
  vector<vector<int>> fourSum(vector<int>& nums, int target) {
  	vector<vector<int>> res;
  	sort(nums.begin(),nums.end());
  	for(int k=0;k<nums.size()-3;k++){
			if(nums[k]>target&&nums[k]>=0) break;
			if(k>0&&nums[k]==nums[k-1]) continue;
			for(int i=k+1;i<nums.size()-2;i++){
				if(nums[k]+nums[i]>target&&nums[i]>=0) break;
				if(i>k+1&&nums[i]==nums[i-1]) continue;
				int left=i+1,right=nums.size()-1;
				while(left<right){
					int n1=nums[left],n2=nums[right];
					if((long)nums[k]+nums[i]+n1+n2==target){
						res.push_back(vector<int>{nums[k],nums[i],n1,n2});
						while(left<right&&nums[left]==n1) left++;
						while(left<right&&nums[right]==n2) right--;
					}else if((long)nums[k]+nums[i]+n1+n2<target) left++;
					else right--;
				}
			}
		}
		return res;
  }
};
